cql3/statements/select_statement.cc

```diff
@@ -689,24 +690,27 @@ indexed_table_select_statement::do_execute_base_query(
             auto command = ::make_lw_shared<query::read_command>(*cmd);
 
             query::result_merger oneshot_merger(cmd->get_row_limit(), query::max_partitions);
-            return utils::result_map_reduce(key_it, key_it_end, [this, &qp, &state, &options, cmd, timeout] (auto& key) {
+            coordinator_result<foreign_ptr<lw_shared_ptr<query::result>>> rresult = co_await utils::result_map_reduce(key_it, key_it_end, [this, &qp, &state, &options, cmd, timeout] (auto& key)
```

Comment:
Don't you need to instantiate all the captured references in the coroutine frame, since this lambda is a coroutine as well? Or, preferably, just not coroutinize this lambda, since it's fairly straight forward code with a single `then` step for launching `query_result()`. This is the kind of issue that has been haunting us for a long time after introducing forward_service, and it was hard to debug because it only occasionally crashed.

Reply:
I should have changed the captures to [&], no need for internal captures.

The change to internal coroutines makes further modification easier, so I'd like to keep it.


Perhaps a safer way to write it is so:

```c++
    rresult = co_await result_map_reduce(key_it, key_it_end, std::ref([&] (...) -> ... {
         // inner coroutine
    }));
```

std:ref means we don't pass a coroutine to result_map_reduce but a reference_wrapper instead. The compiler will keep the lambda coroutine alive until the end of the expression (the last semicolon).